en/lessons/libraries/distillery.md

```diff
@@ -0,0 +1,245 @@
+
+```
+
+Try re-running the application and it should start up successfully. However, you will notice that your migrations to your database have not run. Usually in development we run those migrations manually by calling `mix.ecto migrate`. For the release, we will have to configure it so that it can run the migrations on its own.
+
+
+## Configuring your release
+
+If you need to add further functionality to your release, you can do so by making edits to the `rel/config.exs` file
+
+Often when creating releases, you will find the need to customize your release. This can be done by making edits to the `rel/config.exs` file.
+
+Distillery provides us with the ability to execute code at different points in a release's lifecycle. These points are known as [boot-hooks](https://hexdocs.pm/distillery/1.5.2/boot-hooks.html). The hooks provided by Distillery include
+
+* pre_start
+* post_start
+* pre/post_configure
+* pre/post_stop
+* pre/post_upgrade
+
+
+For our purposes, we're going to be using the `post_start` hook to run our apps migrations in production. Let's first go and create a new release tasks called `migrate`
+
+```
+defmodule BookAppWeb.ReleaseTasks do
+  def migrate do
+    {:ok, _} = Application.ensure_all_started(:book_app)
+
+    path = Application.app_dir(:book_app, "priv/repo/migrations")
+
+    Ecto.Migrator.run(BookApp.Repo, path, :up, all: true)
+  end
+end
+```
+
+*Note* It is good practice to ensure that your applications have all started up properly before running these migrations.
+
+Next, create a new file - `rel/hooks/post_start/migrate.sh` and add the following code:
+
+
+```
+echo "Running migrations"
+
+bin/book_app rpc "Elixir.BookApp.ReleaseTasks.migrate"
+
+```
+
+Finally, in our rel/config.exs file we're going to add the hook to our prod configuration.
+
+Let's replace
+
+```
+environment :prod do
+  set include_erts: true
+  set include_src: false
+  set cookie: :"TkJuF,3nc4)OWPBpPxPDb6mz$>)>a>/v/,l2}W*sUFaz<)bG,v*3pPESE,`XOk{,"
+  set vm_args: "rel/vm.args"
+end
+```
+
+with
+
+```
+environment :prod do
+  set include_erts: true
+  set include_src: false
+  set cookie: :"TkJuF,3nc4)OWPBpPxPDb6mz$>)>a>/v/,l2}W*sUFaz<)bG,v*3pPESE,`XOk{,"
+  set vm_args: "rel/vm.args"
+  set post_start_hooks: "rel/hooks/post_start"
+end
+```
+
+*Note* - This hook only exists in the production release of this application. If we used the default development release it would not run.
+
+In order for this code to run properly, we are using Erlang's `rpc` module which allows us Remote Produce Call service. Basically, this allows us to call a function on a remote node and get the answer. When running in production it is likely that our application will be running in several different nodes
+
+
+Let's take this a step further by adding a new release command. Now that we can run our migrations, we may want to be able to seed our database with information through running a command. First, add a new method to our release tasks.
```

Comment:
what's the difference between something like our release task to migrate the db and a release command to seed the db? i.e. "task" vs. "command"? why do we implement them differently?